services/cronback-scheduler-srv/src/spinner/controller.rs

```diff
@@ -683,6 +689,31 @@ impl SpinnerController {
         Ok(())
     }
 
+    #[tracing::instrument(skip_all, fields(project_id = %context.project_id))]
+    pub async fn delete_project_triggers(
+        &self,
+        context: RequestContext,
+    ) -> Result<(), TriggerError> {
+        let project_id = context.project_id.clone();
+        let triggers = self.triggers.clone();
+
+        // NOTE: This will stall the spinner if the list is too big.
+        tokio::task::spawn_blocking({
+            let project_id = project_id.clone();
+            move || {
+                let mut w = triggers.write().unwrap();
+                w.remove_by_project(&project_id)
+            }
+        })
+        .await?;
+
+        self.store.delete_triggers_by_project(&project_id).await?;
```

Comment:
Don't you think we should update the store first before the in memory state? Updating the store can fail, updating the in memory state won't. So sounds safer to do the store deletion first to avoid inconsistencies?

Reply:
The controller has a process to flush in-memory changes to the database every few seconds, if we delete before holding the write lock on the active triggers map, we risk seeing some triggers getting flushed back to the database before we hold the lock.

It's a tricky bit, and I probably should leave a comment here explaining that delicacy.